### ISSUE TITLE
Audio in Pole Position II distinctly lower quality than Pole Position I

### ISSUE THREAD

**Issue by retrogamedesigner** [libretro/mame2003-plus-libretro#1823]
If I understand correctly, both games should benefit to from the updated sound drivers for various Namco titles. I've tested all versions of both games on a Windows PC, and I'm only hearing the improvement to the first Pole Position. Testing "poleposa" and "poleps2a" on a couple Linux x86_64 PCs, an arm Linux handheld, and an Android handheld, I notice the difference in sound quality between the first and second games as well.

**Comment by arcadez2003** [comment]
Hmm maybe we need some more fixes then.....

0.89u5: Jarek Burczynski improved the Namco 54XX Noise Generator (Type A and B emulated, type C algo still unknown). Removed pp2_17.wav sample.

0.79u1: Nicola Salmoria fixed various mistakes in the polepos ROM loading (polepos2 was using the wrong engine sound ROMs - it's turbo!) and removed various bad dumps from polepos2b and poleps2c, they turned out to be the same version so poleps2c is removed

**Comment by grant2258** [comment]
I remember doing this update years ago not sure what version we are at though

**Comment by arcadez2003** [comment]
Yeah you and me both @grant2258 We're at around MAME80 im looking at MAME90 just now it doesn't seem 
like an awful lot of work but well i've said that before.

**Comment by grant2258** [comment]
haha tooshay should be pretty easy as they say. I miss read that I though the comparison was before that update see that its two different games. On another subject it looks like you have added that sound code properly. The reset only happens when you reset the game like with f3 so I dont think it that, it more than likely going to be needing an updated on our end. I remember doing the ssv soundcore and it slowed things down too much, so im not so keen on touching soundcores anymore haha

**Comment by grant2258** [comment]
Ive have to go pick someone up if you dont get to it ill have a peek tomorrow at least its not a bug in the code we have though. Managed to fix a few things for km  hope  the m8 is a clear run to largs I doubt it though.

**Comment by arcadez2003** [comment]
Safe travels big yin

**Comment by arcadez2003** [comment]
Right from a quick peek at the code.......

From what i can tell driver wise it's galaga.c and polepos.c which is just removing some samples to be replaced
with a new and improved Namco54 soundcore after that a really small change in namcona sound c and h.

So in essence it could be done in a few hours it's just i'd hate to do it only to be told it's not made
the sound better for Pole Position 2 etc etc but it's worth a try at anyrate for a classic game like this one

**Comment by grant2258** [comment]
yea i played this one not sure if it was the first or second though as kid had a accelerator pedal and steering wheel in a stand up cabinet. Checked with google maps junction 16 seems to have a diversion see where I end up haha.

**Comment by arcadez2003** [comment]
Alright i've done galaga.c gaplus.c and polepos.c drivers just need to do the namcona1.c driver and the
two soundcores i think the code should just drop in here i'll doublecheck it tomorrow.

**Comment by retrogamedesigner** [comment]
Just wanted to say thank you to everyone for looking into these issues, as well as for being so responsive in addressing them. You all are the best.

**Comment by arcadez2003** [comment]
Nae worries but keep in mind updating the Namco54 sound might not be the fix we need but fingers crossed

Right so then after looking more closely at the code namcona1 changes are not connected to Namco54 so
having done the drivers last night all that remains is to backport the soundcore changes which all being well
i should have done by the days end.

**Comment by arcadez2003** [comment]
Right @retrogamedesigner the changes are in once we confirm the source will compile with the new code 
you update to the latest build and hopefully the sound will now be better for Pole Position 2.

If not then there are some more fixes which could be backported for the Namco54 sound but that would also
require we also update Namco52 sound and all related drivers which wont be a light in the graft dept so it might
take a while until one of us has the time to work on it.

Regards

**Comment by arcadez2003** [comment]
For my notes if further work should be required......

Bosconian
0.90u3: Derrick Renaud added proper filters to the Namco52 sample player of Bosconian, Galaga, Pole Position and Xevious. Adjusted relative effect volumes per schematics.

0.90u2: Derrick Renaud updated the Namco 54xx noise sound and added R/C values to Bosconian, Galaga, Pole Position and Xevious.

Galaga
0.90u3: Derrick Renaud added proper filters to the Namco52 sample player of Bosconian, Galaga, Pole Position and Xevious. Adjusted relative effect volumes per schematics.

0.90u2: Derrick Renaud updated the Namco 54xx noise sound and added the R/C values to Bosconian, Galaga, Pole Position and Xevious.

Pole Position
0.90u3: Derrick Renaud added proper filters to Pole Position engine sound and proper filters to the Namco52 sample player of Bosconian, Galaga, Pole Position and Xevious. Adjusted relative effect volumes per schematics.

14th January 2005: Derrick Renaud - Added proper filters to engine sound. Also i am working on adding the proper filtering to the speech (Namco 52XX voice effects) of Pole Position. The final sound comes out of pin 14 of the LM324. R/C components and a op-amp affect the filtering.

0.90u2: Derrick Renaud updated the Namco 54XX noise sound and added the R/C values to Bosconian, Galaga, Pole Position and Xevious.

Xevious
0.93: Derrick Renaud swapped Namco 54XX filters on Port A & C. Fixes Xevious sound.

0.90u3: Derrick Renaud added proper filters to the Namco52 sample player of Bosconian, Galaga, Pole Position and Xevious. Adjusted relative effect volumes per schematics.

0.90u2: Derrick Renaud updated the Namco 54XX noise sound and added the R/C values to Bosconian, Galaga, Pole Position and Xevious.

**Comment by arcadez2003** [comment]
@retrogamedesigner Also if you dont mind can you test the following games to make sure they've not got
worse soundwise after this update thanks.......

Bosconian
Galaga
Pole Position
Xevious

**Comment by retrogamedesigner** [comment]
No problem. I'll test those before I update the core, then I'll retest after for comparison.

**Comment by retrogamedesigner** [comment]
That took longer than I intended - I got a little obsessive. I felt like a musician listening to countless takes of a song for the subtle differences, or like many people feel at an optometrist's office. From what I can tell, there was no change to any of the versions of Pole Position 1 or 2. Speaking of those games, I noticed that polepos.zip also shared the lower quality audio of polepos2.zip, polps2a.zip, and poleps2b.zip using build 081f2fc and the older cbfdd3c build of the core. Yes, poleposa.zip and polepos1.zip sounded better than polepos.zip on both builds of the core.

Bosconian, Galaga, and Xevious (all versions of them) all experienced a regression to lower quality audio after the update. I didn't test any others that also use the galaga.c "sourcefile."

**Comment by arcadez2003** [comment]
Oh my this hasn't went well but fear not im updating to use the Namco52 and 54 soundcores from MAME91 this
will be the last roll of the dice TBH

**Comment by grant2258** [comment]
Ill looked at the other soundcore we would need to update the core to the fbneo version or the one that's fixed in mame. The fbneo probably has a re-trigger call already add thats all I can think of. I tend to work from mame so havent compared the two. But they are doing it a different way from the mame fix in the issue.

As far as polepos goes ill update the roms as its the japan sound in the main set. The sets should fit in nicely anyway. 

There would be need to be more detail on what sounds are lower and near in mind sometimes your comparing the new sound calls to samples that where there before. That is the case in xevious 100%.

**Comment by arcadez2003** [comment]
Sure i was thinking new sound handling vs the older samples might bring up a few things have got worse complaints
anyway i've updated to MAME91 just waiting to see if the code slips in here with no build issues.

This really is the last roll of the dice as not long after these commits they added a new CPU to handle the Namco sound
and switched the games over to discrete and well im not prepared to take that on ha ha.

**Comment by grant2258** [comment]
ok ive updated the romsets as well and added a few ones that weren't there. Ill put a pull req in

**Comment by arcadez2003** [comment]
The build is fucked im rolling back to eliminate all of todays commits, this is why i need to stick with small commits
as without my dev setup i cant like the ole days check the code will compile firstly and then test games afterwards

I should really call it day here TBH.

**Comment by arcadez2003** [comment]
> ok ive updated the romsets as well and added a few ones that weren't there. Ill put a pull req in

Aye im rolling back after im done i can merge your commit

**Comment by grant2258** [comment]
leave it ill sort  the compile fixes m8

**Comment by grant2258** [comment]
@retrogamedesigner can you test on mame current. Im guessing your meaning lack of tone in polepos2 whats happening is a filter is on check on mame current both polepos and polepos are filtered there. Is what you mean by lower quality?

**Comment by retrogamedesigner** [comment]
Yes, on the current MAME core both Pole Position 1 and 2 sound identical to me (other than the engine sounds, which I suspect were slightly different from each other on the original arcade hardware). The most obvious difference I was hearing on the 2003plus core between the first and second game was the sound when the "brakes" are used. On the current core, those sound exactly the same.

**Comment by arcadez2003** [comment]
@grant2258 were you able to get those namco sound changes to compile without the need to update streams.c.??

**Comment by grant2258** [comment]
@arcadez2003 just done the update myself seems to be fine with xevious and galaga and pole_pos. @retrogamedesigner  Ive debugged this issue polepos2 seems to be causing some issues with the braking sample being activated with the high  rpm. Simple rename the sample and play without it for a better experience with polepos2.  

This even happens with the 095 test core I have and I have no idea when it was fixed. Ive also update the romsets apart for a few bootlegs to the latest mame sets. Ill add the bootlegs not done at some point.

**Comment by arcadez2003** [comment]
From reading the WIP they first switch to using an actual MCU to handle much of the 54 sound and the samples
are then removed before a few builds later they switch to discrete sounds, after that the rest of the sound is also 
handled via said MCU.

0.112u1: Replaced implementation of Namco 54xx sound chip with new MB8844 CPU core running the original embedded ROM code. Removed old sound core. Pole Position and Bosconian no longer require samples as a result [Nicola Salmora, Ernesto Corvi, Guru]. Added MB88xx (256000 Hz) CPU4 with 1k rom 54xx.bin. Replaced Namco 54XX and Samples sound with 3x DAC sound.

0.131u3: Aaron Giles changed the Namco 52xx and 53xx devices to run the actual MCU instead of simulating the behavior. Hooked them up properly in Pole Position. Added mirrors to Pole Position memory maps. Replaced audio Namco 52XX with 3x MB8843 (1536000 Hz) CPU4/5/6. Changed MB8844 CPU7 clock speed to 1536000 Hz.

So im thinking we've taken it as far as we can as the above would be a ton of work :)

**Comment by arcadez2003** [comment]
@grant2258 the build is failing on a few platforms related to the streams.c changes by looks

**Comment by grant2258** [comment]
naw its something else ill sort it m8.

**Comment by arcadez2003** [comment]
Alrighty bigman i know you'll get it anyway time to stick on my Gammon steaks

**Comment by grant2258** [comment]
ok thats it sorted its picky about where variables being declared at the start of a function.

**Comment by mahoneyt944** [comment]
Is the sound good now?

**Comment by arcadez2003** [comment]
Waiting to hear back from @retrogamedesigner

**Comment by retrogamedesigner** [comment]
Sounds the same to me as before. The brake sounds are very shrill, like a high-pitched whistle, in Pole Position 2. I'm not sure if there's a low-pass filter or something that's only being used on PP1, but the sound effect for the brakes is very different and much richer sounding (more low frequencies to the sound) compared with PP2.

**Comment by mahoneyt944** [comment]
Is it improved when using current mame?

**Comment by retrogamedesigner** [comment]
Oh, yeah. On current mame, both PP1 and PP2 sound identical to each other. I think they both sound great, as I would expect, on current mame. To compare the cores, I'd say what I hear when playing PP1 on mame2003plus is comparable (if not identical) to both games on current mame.

**Comment by retrogamedesigner** [comment]
Just to clarify, I updated the core to bd371b5 before testing. I used the same files to test on 2003plus and current.

**Comment by grant2258** [comment]
@retrogamedesigner the only thing thats different to me is the pitch which might suggest you need another sample for polepos2 for your personal taste. 

I personally use this sample for my testing(ill post below. You could rip the appropriate sample from mame current for each game they could be different pitches causing the issue but im happy enough with the sample I tested with. 

[polepos.zip](https://github.com/user-attachments/files/16675936/polepos.zip)

**Comment by arcadez2003** [comment]
> Oh, yeah. On current mame, both PP1 and PP2 sound identical to each other. I think they both sound great, as I would expect, on current mame. To compare the cores, what I hear when playing PP1 on mame2003plus is comparable (though I'd say not identical) to both games on current mame.

In current MAME they're using the correct dumps of the MCU's that handle the sound and they created new CPU cores
to run them they then switched to use discrete sound with dac interfacing, what we have is basically a simulation of
the sound mcu's with sound samples.

It's chalk and cheese really.

But if there have been no significant improvements for Bosconian Galaga the Pole Position games and xevious 
it might be as well to just roll back to what we had previously.?? as some of the changes to the filtering and streams
core files might negantively have affected other games.

**Comment by mahoneyt944** [comment]
@grant2258 is current mame using a brake sample? Or are they emulating it now?

**Comment by grant2258** [comment]
@mahoneyt944 arcadez already posted the info above https://github.com/libretro/mame2003-plus-libretro/issues/1823#issuecomment-2295293774

**Comment by arcadez2003** [comment]
@grant2258 Yeah that idea had occurred to me also, as you said we could create a seperate sample call for polepos2 
with the sample volume lowered to the preferred level maybe 20 would do it.??

**Comment by grant2258** [comment]
Well @retrogamedesigner is happy with polepos the way it is but it isint perfect but we are using samples. 

Here my take on it somone sampled  pp arcade machine to get this sample and its saved at 22050. This makes it sound very similar I think the brake screech is a bit high but is livable for pp1.

Now we come to pp2 and its sending different data here

https://github.com/libretro/mame2003-plus-libretro/blob/bd371b5c7052e985e37c0f1c9893bdc856683c23/src/sound/namco54.c#L632-L646

so the solution is boot mame into service move a play sample the pp2 brake sample and record it. Now I know this is the issue because polepos1 sounds like pp2 did and pp2 sounds fine now.

So just a different sample interface for pp2 in the machine init with the new sample.

**Comment by retrogamedesigner** [comment]
Hope I'm not causing confusion or interrupting too much. In my opinion, Namco games such as Galaga, Xevious, etc. sounded better before I opened this issue. To be specific, explosions in Galaga and Xevious sound distorted/crackling and other sounds tend to be a bit too bright and chirpy for my taste. I have a couple systems held back at cbfdd3c for that reason.

If it's possible roll back the general changes to the Namco sound drivers (or whatever the specific updates were to investigate this issue) and still look into the samples and how those are handled, I'd say that seems like the way to go.

I tested out the samples @grant2258 posted, preferring the one taken using current MAME. Definitely better overall, and the difference from the first game to the second is slightly less obvious, but still there. For now, I've got my frontends using current MAME for PP1 & 2, but if there is a way to get the samples/audio for both games to match each other on 2003plus, that would be ideal for me.

**Comment by arcadez2003** [comment]
Well as magnus used to say "i've started so i'll finish" im rounding up the code now

**Comment by arcadez2003** [comment]
Alrighty @grant2258 it's all done just as well you mentioned your extra Pole Position romsets as i'd forgot about em
Also for the record guys my mappy changes were not part of the big Namco update and might as well rollback gaplus.c
to use the galaga samples once again.

**Comment by grant2258** [comment]
Well I didn't expect any issues to be honest but im pretty sure it will be volume tweaks and such but like you said people might just get confused whats going on always best to revert and fix in our own time. 

Ill do this the easy way, ill check mame at that update and see if i hear any issues think the old binaries are still available. I appreciate you doing that nooded off earlier been driving what feels like half of scotland today. Sometimes these unpleasing sound changes are actually how the stuff works but ill compare to mame current for a rough idea of what they should sound like.

**Comment by arcadez2003** [comment]
Yeah i know what your saying sometimes people just get used to and prefer how games play in the cores they've 
used the most, i know for a fact alotta people still keep Donkey Kong and Space Invaders using the ole samples
as they dont like how they sound after they were updated to use discrete.

Anyway sometimes the simple fixes are what get's the job done eg if the sample sounds are sometimes too loud 
in Pole Position 2 maybe the user can just use the volume sliders to adjust it to their preference afterall these save 
for everytime they play the game afterwards.

**Comment by grant2258** [comment]
Well ive been looking at things and there are a few issues going on. 

Issue 1 the inputs where broken and this messed up the sound volumes as they are saved in the cfg with the inputs. I had to delete my config files and volumes fixed. People millage may vary depending on what commit they ran at but if you having sound issues the first thing you need to try is deleting your config files.  

Issue 2 the explosion samples volumes needed fixed thankfully that was easy xevious was too low.

issue 3 polepos sample should be 44100 and the reference sample is 22050 thats why its so high pitched. The brakes on polepos1 and polepos2 are the same sound its just polepos2 has a extra range when you leave it in first gear. Ill fix this up later as I think it can be improved a little for polepos2.

@arcadez2003  they took away the volume controls and the game info screens when a game wasn't working so the user cant set them anymore. 

What im going to do is produce some fixes play mame current vs the changes for galaga and xevious then fire them in a zip and you guys can decide if it worse or not. We cant compare it to samples we are using its chalk and cheese.

**Comment by retrogamedesigner** [comment]
From the clips, I'd say all of those sound great on 2003plus.

**Comment by grant2258** [comment]
@retrogamedesigner thats some good news wasn't too much work either thanfully

**Comment by retrogamedesigner** [comment]
I replaced the old brake sample with the one @grant2258 posted from current MAME and updated the core before testing. It sounds amazing. From my point of view, this issue is probably good to close.

Just a couple notes...

1. Unless I'm mistaken, the updated rom list for Pole Position 1 and 2 comes from current Mame, or some recent version of it. I noticed three versions of the games don't load (polepos2bi, polepos2bs, and poleposn), since they are not included in the DAT file. I believe the first two are bootlegs, and in any case, none of those matter to me.

2. Brake samples when playing **polepos** sound identical to **polepos2**, **polepos2a**, **polepos2b**, and **poleposj**. The others, **polelosa1**, **poleposa1n**, and **poleposa2** sound like the brake sample is being played at a slightly lower pitch.

My plan is to to use polepos and polepos2a, since the text is in English and I can use the dipswitch settings to switch from km and km/h to mi and mi/h. They sound identical to me, so I'm happy.

**Comment by grant2258** [comment]
@retrogamedesigner if you add a line of code here just after the else open brace 

https://github.com/libretro/mame2003-plus-libretro/blob/4986bd37b3b2cddcc45b2d88ab4b546c38beca8f/src/sound/namco54.c#L640

if ( data ==120) data =126; 

that should fix the other sets up.

**Comment by retrogamedesigner** [comment]
Thanks. Yes, that's working for me on my Linux desktops. Is that line eventually going to be added and merged in? If so, I'll probably wait on my other platforms, since I have no experience compiling on Windows, Android, or the modified version of Ubuntu 19 that one of my handhelds runs.

**Comment by grant2258** [comment]
Sure I can put a pull request in have no control over if it accepted or not though.

**Comment by mahoneyt944** [comment]
Any further concerns here or can we close 'r ' up?

**Comment by retrogamedesigner** [comment]
As far as I can tell, all is good.
